src/nbl/video/utilities/IUtilities.cpp

```diff
@@ -669,15 +163,662 @@ void IUtilities::updateImageViaStagingBuffer(
+                memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[3] * remainingLayersInRegion;
+            else if (currentBlockInRow == 0 && currentRowInSlice == 0 && currentSliceInLayer > 0)
+            {
+                memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[2] * remainingSlicesInLayer;
+                if(remainingLayersInRegion > 1u)
+                    memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[3] * (remainingLayersInRegion - 1u);
+            }
+            else if (currentBlockInRow == 0 && currentRowInSlice > 0)
+            {
+                memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[1] * remainingRowsInSlice;
+                if(remainingSlicesInLayer > 1u)
+                    memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[2] * (remainingSlicesInLayer - 1u);
+                if(remainingLayersInRegion > 1u)
+                    memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[3] * (remainingLayersInRegion - 1u);
+            }
+            else if (currentBlockInRow > 0)
+            {
+                memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[0] * remainingBlocksInRow;
+                if(remainingRowsInSlice > 1u)
+                    memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[1] * (remainingRowsInSlice - 1u);
+                if(remainingSlicesInLayer > 1u)
+                    memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[2] * (remainingSlicesInLayer - 1u);
+                if(remainingLayersInRegion > 1u)
+                    memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[3] * (remainingLayersInRegion - 1u);
+            }
+        }
+        else
+        {
+            memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[3] * region.imageSubresource.layerCount; // = blockByteSize * imageExtentInBlocks.x * imageExtentInBlocks.y * imageExtentInBlocks.z * region.imageSubresource.layerCount
+        }
+    }
+    return memoryNeededForRemainingRegions;
+}
+
+bool ImageRegionIterator::advanceAndCopyToStagingBuffer(asset::IImage::SBufferCopy& regionToCopyNext, size_t& availableMemory, size_t& stagingBufferOffset, void* stagingBufferPointer)
+{
+    if(isFinished())
+        return false;
+        
+    auto addToCurrentUploadBufferOffset = [&](uint32_t size) -> bool 
+    {
+        const auto initialOffset = stagingBufferOffset;
+        stagingBufferOffset += size;
+        stagingBufferOffset = core::alignUp(stagingBufferOffset, bufferOffsetAlignment);
+        const auto consumedMemory = stagingBufferOffset - initialOffset;
+        if(consumedMemory <= availableMemory)
+        {
+            availableMemory -= consumedMemory;
+            return true;
+        }
+        else
+        {
+            assert(false);
```

Comment:
assert is a program crash, you shouldn't crash the program just cause an aligned up offset doesn't fit!

Reply:
Yea old code, should be removed